Shepherd: Compare specific test names instead of failure counts

## Problem

The shepherd's baseline comparison logic compares failure *counts* between main and worktree, not specific test identifiers. This leads to false positives when different tests fail.

**Example from #2006:**
- Main fails: `test_cli_wrapper_health_routes_to_python` (1 failure)
- Worktree fails: `integration_basic` daemon tests (different failure)
- Shepherd sees "more failures" and blocks, even though they're unrelated

## Current Behavior

From `builder.py` lines 621-631:
```python
# Stage 1: Structured Count Comparison
# If worktree_failures <= baseline_failures → PASS
# If worktree_failures > baseline_failures → FAIL
```

## Proposed Solution

Parse and compare specific test identifiers:
```python
baseline_failures = {"test_cli_wrapper_health"}  
worktree_failures = {"test_cli_wrapper_health", "integration_basic"}
new_failures = worktree_failures - baseline_failures  # {"integration_basic"}
if new_failures:
    fail(f"New test failures: {new_failures}")
```

## Implementation Challenges

- Test name extraction is runner-specific (pytest, cargo, jest, vitest)
- Need parsers for each supported test runner
- Fallback to current behavior if parsing fails

## Acceptance Criteria

- [ ] Extract specific test names from pytest output
- [ ] Extract specific test names from cargo test output
- [ ] Compare test sets instead of counts
- [ ] Only fail if worktree introduces tests that passed on baseline
- [ ] Fallback gracefully if test name parsing fails

## References

- Originated from shepherd reflection on #2006
- Current implementation: `loom-tools/src/loom_tools/shepherd/phases/builder.py`

## Curator Enhancement

### Key Finding: Partial Implementation Already Exists

The builder phase already has `_extract_failing_test_names()` (builder.py lines 572-609) that parses individual test names from pytest, cargo, and vitest output. However, this method is **not used in the primary comparison path**. The comparison flow currently is:

1. **Stage 1** (`_compare_test_results`, line 611): Compares integer failure counts via `_parse_failure_count()`
2. **Stage 2** (fallback, line 886): Compares normalized error lines via `_extract_error_lines()`

Neither stage uses the already-extracted test names for set-based comparison.

### Implementation Guidance

**Recommended approach — Add Stage 1.5 between existing stages:**

1. When Stage 1 detects `worktree_count > baseline_count`, don't immediately return `True`
2. Instead, call `_extract_failing_test_names()` on both outputs
3. Compute `new_failures = worktree_names - baseline_names`
4. If `new_failures` is empty (same tests failing, count discrepancy likely from flaky output), return `None` (pre-existing)
5. If `new_failures` is non-empty, return `True` with the specific new test names for reporting
6. If either set is empty (parsing failed), fall through to existing Stage 2

**Why this approach:**
- Reuses existing `_extract_failing_test_names()` which already handles pytest, cargo, and vitest
- Minimal code changes — adds a refinement step rather than replacing the existing logic
- Graceful degradation — if name extraction fails for either side, falls back to current behavior
- The existing Stage 2 line-based comparison remains as final fallback

**Specific code location:** Modify `_compare_test_results()` at builder.py lines 625-631. Currently:
```python
if baseline_count is not None and worktree_count is not None:
    if worktree_count <= baseline_count:
        return None  # Pre-existing failures
    return True      # New failures detected
```

Change to:
```python
if baseline_count is not None and worktree_count is not None:
    if worktree_count <= baseline_count:
        return None  # Pre-existing failures
    # Stage 1.5: Refine with test name comparison
    baseline_names = set(self._extract_failing_test_names(baseline_output))
    worktree_names = set(self._extract_failing_test_names(worktree_output))
    if baseline_names and worktree_names:
        new_failures = worktree_names - baseline_names
        if not new_failures:
            return None  # Same tests failing
        # Log the specific new failures for reporting
        ...
        return True
    # Fall through to Stage 2 if name extraction fails
    return True  # Conservative: if can't parse names, trust counts
```

**Note on `_extract_failing_test_names()` coverage:**
- **Pytest**: Handles `FAILED path::test_name` format (line 581)
- **Cargo**: Handles `test path ... FAILED` format (line 589)  
- **Vitest/Jest**: Handles `FAIL path` format (line 597)
- Returns empty list if no patterns match — this is the graceful fallback trigger

### Edge Cases to Handle

1. **Flaky tests**: Test A fails on baseline, Test B fails on worktree — these are genuinely new failures and should be flagged
2. **Subset failures**: Baseline has {A, B} failing, worktree has {A} — this is an improvement, current `<=` count check handles this
3. **Name parsing mismatch**: If one runner's output is parseable and the other isn't, fall back to count comparison
4. **Multi-pipeline**: When multiple test runners run in sequence (e.g., cargo then vitest), test names from different runners won't collide since they use different formats

## Affected Files

- `loom-tools/src/loom_tools/shepherd/phases/builder.py` — Modify `_compare_test_results()` (lines 611-639) to add Stage 1.5 name-based comparison; potentially improve `_extract_failing_test_names()` (lines 572-609) if additional patterns are needed
- `loom-tools/tests/shepherd/test_phases.py` — Add tests to `TestBuilderCompareTestResults` class (line 2260+) for name-based comparison scenarios; extend `TestBuilderExtractFailingTestNames` (line 2214+) with edge cases

## Test Plan

- [ ] **Unit tests for name-based comparison**: Add cases where counts differ but test names overlap (should pass), and cases where new test names appear (should fail)
- [ ] **Regression test for #2006 scenario**: Baseline fails `test_cli_wrapper_health`, worktree fails `integration_basic` — should detect `integration_basic` as a new failure
- [ ] **Fallback test**: When `_extract_failing_test_names()` returns empty for one side, verify fallback to count-based or Stage 2 behavior
- [ ] **Multi-runner test**: Pipeline with pytest + vitest output, verify test names from different runners don't collide
- [ ] **Manual verification**: Run shepherd on a worktree with known pre-existing failures and confirm it no longer false-positives
- [ ] **Edge case — flaky test swap**: Baseline fails {A}, worktree fails {B} — should flag B as new failure even though count is same

--- a/loom-tools/src/loom_tools/shepherd/phases/builder.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/builder.py
@@ -627,7 +627,20 @@ class BuilderPhase:
                 # Worktree has same or fewer failures — pre-existing
                 return None
 
-            # Worktree has more failures — new regressions
+            # Stage 1.5: Refine with test name comparison
+            # When worktree has more failures, check if they're genuinely new
+            # tests or just flaky count discrepancies
+            baseline_names = self._extract_failing_test_names(baseline_output)
+            worktree_names = self._extract_failing_test_names(worktree_output)
+            if baseline_names and worktree_names:
+                new_failures = worktree_names - baseline_names
+                if not new_failures:
+                    # Same tests failing — count discrepancy is noise
+                    return None
+                # Genuinely new test failures detected
+                return True
+
+            # Name extraction failed for at least one side — trust counts
             return True
 
         # If only one side parsed, we can't do structured comparison

--- a/loom-tools/tests/shepherd/test_phases.py
+++ b/loom-tools/tests/shepherd/test_phases.py
@@ -2341,6 +2341,119 @@ class TestBuilderCompareTestResults:
         worktree = "Tests  1 failed, 1074 passed\n"
         assert builder._compare_test_results(baseline, worktree) is True
 
+    def test_higher_count_but_same_test_names_returns_none(self) -> None:
+        """Higher count but identical test names -> count discrepancy is noise."""
+        builder = BuilderPhase()
+        baseline = (
+            "FAILED tests/test_foo.py::test_bar - AssertionError\n"
+            "========================= 1 failed, 14 passed in 2.45s ========================\n"
+        )
+        worktree = (
+            "FAILED tests/test_foo.py::test_bar - AssertionError\n"
+            "========================= 2 failed, 13 passed in 2.51s ========================\n"
+        )
+        assert builder._compare_test_results(baseline, worktree) is None
+
+    def test_higher_count_with_new_test_name_returns_true(self) -> None:
+        """Higher count with genuinely new test name -> new failures."""
+        builder = BuilderPhase()
+        baseline = (
+            "FAILED tests/test_foo.py::test_bar - AssertionError\n"
+            "========================= 1 failed, 14 passed in 2.45s ========================\n"
+        )
+        worktree = (
+            "FAILED tests/test_foo.py::test_bar - AssertionError\n"
+            "FAILED tests/test_baz.py::test_qux - ValueError\n"
+            "========================= 2 failed, 13 passed in 2.51s ========================\n"
+        )
+        assert builder._compare_test_results(baseline, worktree) is True
+
+    def test_regression_2006_different_test_names(self) -> None:
+        """Regression test for #2006: different tests failing should be detected.
+
+        Baseline fails test_cli_wrapper_health, worktree fails integration_basic.
+        The name-based comparison should detect integration_basic as new.
+        """
+        builder = BuilderPhase()
+        baseline = (
+            "test utils::tests::test_cli_wrapper_health ... FAILED\n"
+            "test result: FAILED. 8 passed; 1 failed; 0 ignored\n"
+        )
+        worktree = (
+            "test utils::tests::test_cli_wrapper_health ... FAILED\n"
+            "test integration::tests::integration_basic ... FAILED\n"
+            "test result: FAILED. 7 passed; 2 failed; 0 ignored\n"
+        )
+        assert builder._compare_test_results(baseline, worktree) is True
+
+    def test_flaky_test_swap_same_count(self) -> None:
+        """Flaky test swap: baseline {A}, worktree {B} with same count.
+
+        Count comparison says <= so returns None (pre-existing). This is
+        correct behavior — same count means no increase in failures.
+        The name-based refinement only triggers when worktree_count > baseline_count.
+        """
+        builder = BuilderPhase()
+        baseline = (
+            "FAILED tests/test_a.py::test_alpha - Error\n"
+            "========================= 1 failed, 14 passed in 2.45s ========================\n"
+        )
+        worktree = (
+            "FAILED tests/test_b.py::test_beta - Error\n"
+            "========================= 1 failed, 14 passed in 2.51s ========================\n"
+        )
+        # Same count -> returns None via the <= check (before name comparison)
+        assert builder._compare_test_results(baseline, worktree) is None
+
+    def test_name_extraction_fails_one_side_trusts_counts(self) -> None:
+        """When name extraction fails for one side, fall back to count comparison."""
+        builder = BuilderPhase()
+        # Baseline has no parseable test names, just a summary
+        baseline = (
+            "========================= 1 failed, 14 passed in 2.45s ========================\n"
+        )
+        # Worktree has parseable test names
+        worktree = (
+            "FAILED tests/test_foo.py::test_bar - Error\n"
+            "FAILED tests/test_baz.py::test_qux - Error\n"
+            "========================= 2 failed, 13 passed in 2.51s ========================\n"
+        )
+        # Baseline names empty -> can't do name comparison -> trusts counts
+        assert builder._compare_test_results(baseline, worktree) is True
+
+    def test_name_extraction_fails_both_sides_trusts_counts(self) -> None:
+        """When name extraction fails for both sides, fall back to count comparison."""
+        builder = BuilderPhase()
+        baseline = (
+            "========================= 1 failed, 14 passed in 2.45s ========================\n"
+        )
+        worktree = (
+            "========================= 2 failed, 13 passed in 2.51s ========================\n"
+        )
+        # Neither side has parseable names -> trusts counts -> True
+        assert builder._compare_test_results(baseline, worktree) is True
+
+    def test_multi_runner_pytest_and_vitest(self) -> None:
+        """Multi-runner output: pytest + vitest names don't collide."""
+        builder = BuilderPhase()
+        baseline = (
+            "FAILED tests/test_foo.py::test_bar - AssertionError\n"
+            "========================= 1 failed, 4 passed in 1.2s ========================\n"
+            "FAIL src/foo.test.ts\n"
+            " Tests  1 failed, 10 passed\n"
+        )
+        worktree = (
+            "FAILED tests/test_foo.py::test_bar - AssertionError\n"
+            "========================= 1 failed, 4 passed in 1.3s ========================\n"
+            "FAIL src/foo.test.ts\n"
+            "FAIL src/bar.test.ts\n"
+            " Tests  2 failed, 9 passed\n"
+        )
+        # Baseline: 2 total failures (1 pytest + 1 vitest)
+        # Worktree: 3 total failures (1 pytest + 2 vitest)
+        # New failure: src/bar.test.ts
+        assert builder._compare_test_results(baseline, worktree) is True
+
 
 class TestBuilderFallbackComparison:
     """Test that fallback to line-based comparison works when parsing fails."""

